platform-sdk/docs/proposals/metric-labels/README.md

```diff
@@ -0,0 +1,175 @@
+    // void remove(@NonNull String category, @NonNull String name);
+    
+    @Nullable
+    Object getValue(@NonNull String category, @NonNull String name, @NonNull Set<Label> labels);
+    
+    void remove(@NonNull String category, @NonNull String name, @NonNull Set<Label> labels);
+}
+```
+
+Since the old methods are only used in tests or demos today we can do that change easily.
+
+#### Snapshot API
+
+The snapshot API will be extended to return the labels of a metric for a snapshot.
+By doing so any monitoring system that uses the snapshot API can use the labels.
+Today we only have 2 system that uses the snapshot API: the prometheus endpoint and the CSV exporter.
+Since no class of the snapshot API is public today (based on the internal usage of `nodeId`) we plan to move parts
+of the snapshot API to the public API.
+
+### Migration
+
+Next to extending the public API we need to migrate the current metrics to the new concept.
+
+#### Migration of the `category`, `name`, and `nodeId`
+
+The `category`, `name`, and `nodeId` information of a metric will be migrated to labels.
+The key of a metric is the combination of the `category` and `name`.
+That will not be changed but next to that both values will be added as labels to the metric.
+By doing so a metric can be identified by its key and labels.
+
+Let's assume a metric with the category `myCategory`, the name `myName`.
+As today the key of the metric is `myCategory.myName`.
+Next to that the metric will have 2 labels: `category=myCategory` and `name=myName`.
+The unique identifier of the metric is the key and the labels.
+That is a breaking change since today the key is the unique identifier of a metric.
+Since we do not have api today to add labels the migration should be minimalistic.
+
+#### Migration of the prometheus endpoint
+
+The `nodeId` of a platform is already added as a label today when using the prometheus endpoint.
+For that endpoint we just need to remove the hard coded `nodeId` label and add a general algorithm to add labels.
+
+#### Migration of the CSV exporter
+
+The CSV exporter use the `nodeId` today to create the file name.
+Since the CSV exporter is only used by the platform module we can continue with that approach.
+
+## Testing
+
+The metrics module is tested by unit tests today. We need to add tests for the new concept of labels.
+Next to that we need to add tests of the CSV exporter and the prometheus endpoint to ensure that the labels are added correctly.
+Here we need to check that a platform metric (that has a `nodeId` label) is still working correctly and ends in the exact same output as today.
+
+## Open Questions
+
+Please help me to answer the following questions before we can start with the implementation:
+
+- Should the `getLabels()` method return `Set`, `Collection` or `List`?
+- Should we add a new methods to the `Metrics` interface to get for example all metrics with a specific label?
+- How will a label with a `null` value be handled?
```

Comment:
>- Should the `getLabels()` method return `Set`, `Collection` or `List`?

Set. Order is not important. Repeated labels are not desired. 

> - Should we add a new methods to the `Metrics` interface to get for example all metrics with a specific label?

As I don't agree that it should be part of metric creation, I don't agree with this functionality either. 

> - How will a label with a `null` value be handled?

Labels should be optional; a label with a null value should be similar to not being indicated.